omegaconf/basecontainer.py

```diff
@@ -194,7 +223,16 @@ def convert(val: Node) -> Any:
             assert isinstance(inter, str)
             return inter
         elif conf._is_missing():
-            return MISSING
+            if throw_on_missing:
+                conf._format_and_raise(
+                    key=None,
+                    value=None,
+                    cause=MissingMandatoryValue(
+                        f"Encountered a missing {type(conf).__name__} with `throw_on_missing==True`"
```

Comment:
Yes, they were different. I made the change in https://github.com/omry/omegaconf/pull/730/commits/52d5c89f7cdd31575128e6943f15d8db7942727c.

Reply:
Actually, looking at it more closely, I realize now that this exception can only be raised for the config `_to_container()` is  called on. If there is a missing container (e.g., `DictConfig("???")`) inside the the config, this will be caught before we reach this point.

As a result, I'm not sure anymore that my comment was relevant, sorry. I actually wonder if we really want to support this use case, because in general we want something like this to be true: 
```
OmegaConf.create(OmegaConf.to_container(cfg, throw_on_missing=False)) == cfg
```
But this is not true if `cfg` is `DictConfig("???")`.

As a result, it might be better to always make it an error. But it'd be good to have @omry's opinion on it.

Reply:
Let's discuss this on #727. See my last comment there.

As I said there, to_container is losing type safety. Missing containers are used in conjunction with Structured Configs (when you have a missing SC field):
```python
@dataclass
class Config:
  user: User = MISSING # DictConfig("???") with a backing User type.
```

Reply:
> Missing containers are used in conjunction with Structured Configs (when you have a missing SC field):
> 
> ```python
> @dataclass
> class Config:
>   user: User = MISSING # DictConfig("???") with a backing User type.
> ```

Yes but in your example this situation still won't occur because:
- `to_container(cfg, throw_on_missing=True)` would raise an exception when we detect that `cfg.user` is missing (this check is done before we get a chance to call `to_container(cfg._get_node("user"))`)
- `to_container(cfg.user, throw_on_missing=True)` would raise an exception before `to_container()` is called (trying to access `cfg.user` will fail).

As far as I can tell this error can only be triggered if a user ends up manipulating directly a `DictConfig("???")`, but I'm not sure it's something that is supposed to happen in practice. Which is why I was thinking that maybe we shouldn't even try to support it.

Reply:
> As far as I can tell this error can only be triggered if a user ends up manipulating directly a DictConfig("???"), but I'm not sure it's something that is supposed to happen in practice. Which is why I was thinking that maybe we shouldn't even try to support it.

I agree. and this is not a part of the public API.
In fact, DictConfig itself is not meant to be instantiated directly at all. The documented way to create configs is OmegaConf.create().

Reply:
So, actually, looking at it more closely, this situation (a `DictConfig("???")` being processed by `to_container()`) can occur with structured configs, because the `to_container()` implementation takes a different code path (compared to non-structured configs).
As a result, we should indeed support it => resolving this comment.